libraries/AP_LocationDB/AP_LocationDB.cpp

```diff
@@ -0,0 +1,414 @@
+             _size -= 1; // decrease the size of database by one
+
+            if (_size != 0) {
+                // bring the last element at the place of item to be removed
+                // the items with index equal to or beyond database size are not considered to be a part of the database
+                _items[item_idx] = _items[_size];
+            }
+
+            return true;
+        }
+    }
+
+    return false;
+}
+
+// return true if we successfully update the item at given index of location database
+bool AP_LocationDB::update_item_at_index(const uint16_t index, AP_LocationDB_Item &new_item) {
+    // check of database is healthy and the key for the new item is valid
+    // also check if the passed index falls inside the current database size
+    if (!healthy() || !is_valid_key(new_item.get_key()) || index >= _size) {
+        return false;
+    }
+
+    {
+        WITH_SEMAPHORE(db_sem);
+        // copy new item at the place of the old item
+        _items[index] = new_item;
+    }
+
+    return true;
+}
+
+bool AP_LocationDB::update_item(const uint32_t key, AP_LocationDB_Item &new_item) {
+    if (!healthy() || !is_valid_key(key)) {
+        return false;
+    }
+
+#if !APM_BUILD_TYPE(APM_BUILD_UNKNOWN) && AP_AHRS_ENABLED
+    Vector3f item_pos;
+    Vector3f our_pos;
+    if (new_item.get_pos_cm_NEU(item_pos) && AP::ahrs().get_relative_position_NED_origin(our_pos)) {
+        our_pos.z = -our_pos.z;
+        item_pos = item_pos / 100; // cm to m
+        // check if the items is inside the inclusion radius
+        if ((item_pos - our_pos).length_squared() > (_inc_radius.get() * _inc_radius.get())) {
+            // item very far away from the vehicle
+            return false; 
+        }
+    }
+#endif
+
+    {
+        WITH_SEMAPHORE(db_sem);
+        uint16_t item_idx;
+        // get index of item to be updated
+        // update item if the index is successfully retrieved
+        if (get_item_index(key, item_idx)) {
+            return update_item_at_index(item_idx, new_item);
+        }
+    }
+
+    return false;
+}
+
+bool AP_LocationDB::is_valid_key(uint32_t key) {
+    const KeyDomain key_domain = static_cast<KeyDomain>(key >> 24);
+
+    // domain specific checks, if any
+    switch (key_domain) {
+        case KeyDomain::MAVLINK:
+            return (key & 255U) != 255U;  // last byte should not be 255, this means the mavlink msgid is not supported
+        case KeyDomain::ADSB:
+        case KeyDomain::SCRIPTING:
+            return true;
+    }
+
+    // we reach here only if the key is for an invalid domain
+    return false;
+}
+
+// helper method to construct location database key for a mavlink item
+uint32_t AP_LocationDB::construct_key_mavlink(uint8_t sysid, uint8_t compid, uint8_t msgid) {
+    return (uint8_t)KeyDomain::MAVLINK << 24 | sysid << 16 | compid << 8 | short_mav_msg_id(msgid);
+}
+
+// helper method to construct location database key for an adsb item
```

Comment:
```suggestion
#if AP_LOCATIONDB_KEYDOMAIN_ADSB_ENABLED
// helper method to construct location database key for an adsb item
```

... and in the configuration file you would
```
#ifndef AP_LOCATIONDB_KEYDOMAIN_ADSB_ENABLED
#define AP_LOCATIONDB_KEYDOMAIN_ADSB_ENABLED HAL_ADSB_ENABLED
#24377 
```

(and include `AP_ADSB_config.h` in `AP_LocationDB_config.h`)

Similarly for the other domains.

Reply:
Done for ADSB and SCRIPTING domains. Thanks. I couldn't find right define for MAVLINK domain though.